contracts/marketpools.js

```diff
@@ -0,0 +1,350 @@
+  if (!api.assert(api.BigNumber(api.BigNumber(quoteDelta).dividedBy(baseDelta).toFixed(pool.precision, api.BigNumber.ROUND_DOWN)).eq(p),
+    `constant price ${api.BigNumber(quoteDelta).dividedBy(baseDelta)}, expected ${p}`)) return false;
+  return true;
+}
+
+function validateSwap(pool, baseDelta, quoteDelta, maxSlippage) {
+  const k = api.BigNumber(pool.baseQuantity).times(pool.quoteQuantity).toFixed(pool.precision, api.BigNumber.ROUND_HALF_UP);
+  const baseAdjusted = api.BigNumber(pool.baseQuantity).plus(baseDelta);
+  const quoteAdjusted = api.BigNumber(pool.quoteQuantity).plus(quoteDelta);
+  const p = api.BigNumber(pool.quoteQuantity).dividedBy(pool.baseQuantity).toFixed(pool.precision, api.BigNumber.ROUND_DOWN);
+  const pAdjusted = api.BigNumber(quoteAdjusted).dividedBy(baseAdjusted).toFixed(pool.precision, api.BigNumber.ROUND_DOWN);
+  const slippage = api.BigNumber(pAdjusted).minus(p).abs().dividedBy(p);
+  if (!api.assert(api.BigNumber(slippage).lte(maxSlippage), 'exceeded max slippage for swap')) return false;
+  if (!api.assert(api.BigNumber(api.BigNumber(baseAdjusted).times(quoteAdjusted).toFixed(pool.precision, api.BigNumber.ROUND_HALF_UP)).eq(k),
+    `constant product ${api.BigNumber(baseAdjusted).times(quoteAdjusted)}, expected ${k}`)) return false;
+  return true;
+}
+
+async function validateTokenPair(tokenPair) {
+  if (!api.assert(typeof (tokenPair) === 'string' && tokenPair.indexOf(':') !== -1, 'invalid tokenPair format')) return false;
+  const [baseSymbol, quoteSymbol] = tokenPair.split(':');
+  if (!api.assert(baseSymbol !== quoteSymbol, 'tokenPair cannot be the same token')
+    || !api.assert(await api.db.findOneInTable('tokens', 'tokens', { symbol: baseSymbol }), 'baseSymbol does not exist')
+    || !api.assert(await api.db.findOneInTable('tokens', 'tokens', { symbol: quoteSymbol }), 'quoteSymbol does not exist')) {
+    return false;
+  }
+  return true;
+}
+
+async function validatePool(tokenPair) {
+  const [baseSymbol, quoteSymbol] = tokenPair.split(':');
+  const pool = await api.db.findOne('pools', { tokenPair });
+  const revPool = await api.db.findOne('pools', { tokenPair: [quoteSymbol, baseSymbol].join(':') });
+  if (!api.assert(pool === null && revPool === null, 'a pool already exists for this tokenPair')) return false;
+  return true;
+}
+
+async function updatePoolStats(pool, baseAdjusted, quoteAdjusted, swap = true) {
+  const uPool = pool;
+  // precise quantities are needed here for K calculation
+  // remainder are statistical and can be rounded (updated for swaps only)
+  uPool.baseQuantity = api.BigNumber(pool.baseQuantity).plus(baseAdjusted);
+  uPool.quoteQuantity = api.BigNumber(pool.quoteQuantity).plus(quoteAdjusted);
+
+  // if all LP is removed, don't update the last price
+  if (uPool.baseQuantity.gt(0) && uPool.quoteQuantity.gt(0)) {
+    uPool.basePrice = api.BigNumber(uPool.quoteQuantity).dividedBy(uPool.baseQuantity).toFixed(pool.precision, api.BigNumber.ROUND_DOWN);
```

Comment:
These are for stats / display only, so rounding or not probably doesn't matter? Or these might be used for oracle purposes right? So maybe rounding would be consistent?

Reply:
These two prices are not currently used for anything in the contract (display only)